native_client/deepspeech.h

```diff
@@ -1,137 +1,172 @@
+
+struct StreamingState;
+
+/**
+ * @brief An object providing an interface to a trained DeepSpeech model.
+ *
+ * @param aModelPath The path to the frozen model graph.
+ * @param aNCep The number of cepstrum the model was trained with.
+ * @param aNContext The context window the model was trained with.
+ * @param aAlphabetConfigPath The path to the configuration file specifying
+ *                            the alphabet used by the network. See alphabet.h.
+ * @param aBeamWidth The beam width used by the decoder. A larger beam
+ *                   width generates better results at the cost of decoding
+ *                   time.
+ * @param[out] retval a ModelState pointer
+ *
+ * @return Zero on success, non-zero on failure.
+ */
+DEEPSPEECH_EXPORT
+int DS_CreateModel(char* aModelPath,
+                   int aNCep,
+                   int aNContext,
+                   char* aAlphabetConfigPath,
+                   int aBeamWidth,
+                   ModelState** retval);
+
+/**
+ * @brief Frees associated resources and destroys model object.
+ */
+DEEPSPEECH_EXPORT
+void DS_DestroyModel(ModelState* ctx);
+
+/**
+ * @brief Enable decoding using beam scoring with a KenLM language model.
+ *
+ * @param aCtx The ModelState pointer for the model being changed.
+ * @param aAlphabetConfigPath The path to the configuration file specifying
+ *                            the alphabet used by the network. See alphabet.h.
+ * @param aLMPath The path to the language model binary file.
+ * @param aTriePath The path to the trie file build from the same vocabu-
+ *                  lary as the language model binary.
+ * @param aLMWeight The weight to give to language model results when sco-
+ *                  ring.
+ * @param aValidWordCountWeight The weight (bonus) to give to beams when
+ *                              adding a new valid word to the decoding.
+ */
+DEEPSPEECH_EXPORT
+void DS_EnableDecoderWithLM(ModelState* aCtx,
+                            char* aAlphabetConfigPath,
+                            char* aLMPath,
+                            char* aTriePath,
+                            float aLMWeight,
+                            float aValidWordCountWeight);
+
+/**
+ * @brief Use the DeepSpeech model to perform Speech-To-Text.
+ *
+ * @param aCtx The ModelState pointer for the model to use.
+ * @param aBuffer A 16-bit, mono raw audio signal at the appropriate
+ *                sample rate.
+ * @param aBufferSize The number of samples in the audio signal.
+ * @param aSampleRate The sample-rate of the audio signal.
+ *
+ * @return The STT result. The user is responsible for freeing the string.
+ */
+DEEPSPEECH_EXPORT
+char* DS_SpeechToText(ModelState* aCtx,
+                      short* aBuffer,
+                      unsigned int aBufferSize,
+                      int aSampleRate);
+
+/**
+ * @brief Setup a context used for performing streaming inference.
+ *        the context pointer returned by this function can then be passed
```

Comment:
The struct holds the streaming state, and it's used as a context pointer. We don't use it everywhere because you can compute several recognitions for a single `ModelState`. We wouldn't want to e.g. load the model from disk on every STT call.

Reply:
My only concern here is with the naming.

The name "context pointer" is a bit ambiguous, it could also refer to the `ModelState`, as that is also a context pointer. However, the name "streaming state" is less ambiguous.